et_quantize.py

```diff
@@ -0,0 +1,473 @@
+):
+    # TODO: remove this
+    scales, zero_points = unpack_scales_and_zeros(scales_and_zeros)
+    n_bit = 4
+    quant_min = -(2 ** (n_bit - 1))
+    quant_max = 2 ** (n_bit - 1) - 1
+    return torch.ops.quantized_decomposed.dequantize_per_channel_group(
+        w_int8, scales, zero_points, quant_min, quant_max, torch.int8, group_size, precision
+    )
+
+
+def down_size(size):
+    assert size[-1] % 2 == 0, f"{size} last dim not divisible by two"
+    return (*size[:-1], size[-1] // 2)
+
+
+def up_size(size):
+    return (*size[:-1], size[-1] * 2)
+
+
+quantized_decomposed_lib.define("pack_int4_from_int8(Tensor int8_data) -> Tensor")
+
+
+@impl(quantized_decomposed_lib, "pack_int4_from_int8", "CompositeExplicitAutograd")
+def pack_int4_from_int8(int8_data: torch.Tensor) -> torch.Tensor:
+    # converting to uint8 for operations
+    shape = int8_data.shape
+    assert shape[-1] % 2 == 0
+    int8_data = int8_data.contiguous().view(-1)
+    return (int8_data[::2] << 4 | int8_data[1::2]).view(down_size(shape))
+
+
+quantized_decomposed_lib.define("unpack_int4_to_int8(Tensor int8_data) -> Tensor")
+
+
+@impl(quantized_decomposed_lib, "unpack_int4_to_int8", "CompositeExplicitAutograd")
+def unpack_int4_to_int8(int8_data: torch.Tensor) -> torch.Tensor:
+    """Get the original weight from the normalized float weight format"""
+    # since we are using int8 we will decode 2 entries per byte
+    # Shift elements down 4 and select out the bottom 4 bits
+    shape = int8_data.shape
+    first_elements = (int8_data >> 4).to(torch.int8)
+    second_elements = (int8_data & 0b1111).to(torch.int8)
+    return torch.stack([first_elements, second_elements], dim=-1).view(up_size(shape))
+
+################ These will be moved to pytorch or torchao later ##############
+
+def per_token_dynamic_quant(input: torch.Tensor) -> torch.Tensor:
+    orig_dtype = input.dtype
+    (
+        scales,
+        zero_points,
+    ) = torch.ops.quantized_decomposed.choose_qparams_per_token(input, torch.int8)
+
+    # TODO: get these from torch.int8
+    quant_min = -128
+    quant_max = 127
+    input = torch.ops.quantized_decomposed.quantize_per_token(
+        input, scales, zero_points, quant_min, quant_max, torch.int8
+    )
+    input = torch.ops.quantized_decomposed.dequantize_per_token(
+        input, scales, zero_points, quant_min, quant_max, torch.int8, orig_dtype
+    )
+    return input
+
+def linear_forward_8da4w(x, weight_int8, scales_and_zeros, out_features, groupsize, precision):
+    """8 bit per token dynamic quantization for activation and 4 bit per channel group quantization
+    for weight
+    """
+    x = per_token_dynamic_quant(x)
+    origin_x_size = x.size()
+    x = x.reshape(-1, origin_x_size[-1])
+    scales_and_zeros = scales_and_zeros.to(torch.float)
+    w_dq = group_dequantize_tensor_symmetric(weight_int8, scales_and_zeros, groupsize, precision)
+    x = x.to(precision)
+    c = torch.ops.aten.linear(x, w_dq)
```

Comment:
For fp16, we will ahve to make some changes here that may result in the need of a different pattern in xnnpack quantizer